File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque

m,n = map(int,input().split())

tomato_box = []

for i in range(n):
    tomato_box.append(list(map(int, input().strip().split())))

# 상하좌우 방향
dx = [-1,1,0,0]
dy = [0,0,-1,1]

cnt = 0

# 값이 1인 토마토 모두 큐에 넣기
queue = deque()
for i in range(n):
  for j in range(m):
    if tomato_box[i][j] == 1:
      queue.append((i,j))


while queue:
  x,y = queue.popleft()
  for i in range(4):
    nx = x + dx[i]
    ny = y + dy[i]

    if nx >= 0 and nx < n and ny >= 0 and ny < m:
      if tomato_box[nx][ny] == 0:
        tomato_box[nx][ny] = tomato_box[x][y] + 1
        queue.append((nx, ny))

for i in tomato_box:
  for j in i:
    if j == 0:
      print(-1)
      exit(0)
  cnt = max(cnt, max(i)) # 이전의 cnt 값과 max(i) 중 더 큰 값

print(cnt-1)

File: 백준/Gold/7576. 토마토/토마토.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
rl.on("line", (line) => {
  input.push(line.trim());
}).on("close", () => {
  const [M, N] = input[0].split(" ").map(Number);
  const board = input.slice(1).map((line) => line.split(" ").map(Number));

  const dx = [1, 0, -1, 0];
  const dy = [0, 1, 0, -1];
  const queue = [];

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (board[i][j] === 1) queue.push([i, j]);
    }
  }

  let front = 0;
  while (front < queue.length) {
    const [x, y] = queue[front++];

    for (let i = 0; i < 4; i++) {
      const nx = x + dx[i];
      const ny = y + dy[i];

      if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
      if (board[nx][ny] !== 0) continue;

      board[nx][ny] = board[x][y] + 1;
      queue.push([nx, ny]);
    }
  }

  let max = 0;
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (board[i][j] === 0) {
        console.log(-1);
        return;
      }
      max = Math.max(max, board[i][j]);
    }
  }

  console.log(max - 1);
});
